packages/core/admin/admin/src/utils/baseQuery.ts

```diff
@@ -19,56 +20,90 @@ interface UnknownApiError {
 
 type BaseQueryError = ApiError | UnknownApiError;
 
+let refreshPromise: Promise<string> | null = null;
+
+const isAuthPath = (url: string) => /^\/admin\/(login|logout|access-token)\b/.test(url);
+
 const simpleQuery: BaseQueryFn<string | QueryArguments, unknown, BaseQueryError> = async (
   query,
-  { signal }
+  api
 ) => {
-  try {
-    const { get, post, del, put } = getFetchClient();
+  const { signal, dispatch } = api as { signal?: AbortSignal; dispatch: (a: any) => void };
 
-    if (typeof query === 'string') {
-      const result = await get(query, { signal });
-      return { data: result.data };
-    } else {
-      const { url, method = 'GET', data, config } = query;
-
-      if (method === 'POST') {
-        const result = await post(url, data, {
-          ...config,
-          signal,
-        });
-        return { data: result.data };
-      }
-
-      if (method === 'DELETE') {
-        const result = await del(url, {
-          ...config,
-          signal,
-        });
-        return { data: result.data };
-      }
-
-      if (method === 'PUT') {
-        const result = await put(url, data, {
-          ...config,
-          signal,
-        });
-        return { data: result.data };
-      }
+  const executeQuery = async (queryToExecute: string | QueryArguments) => {
+    const { get, post, del, put } = getFetchClient();
+    if (typeof queryToExecute === 'string') {
+      const result = await get(queryToExecute, { signal });
+      return result;
+    }
 
-      /**
-       * Default is GET.
-       */
-      const result = await get(url, {
-        ...config,
-        signal,
-      });
-      return { data: result.data };
+    const { url, method = 'GET', data, config } = queryToExecute;
+    if (method === 'POST') {
+      return post(url, data, { ...config, signal });
+    }
+    if (method === 'DELETE') {
+      return del(url, { ...config, signal });
+    }
+    if (method === 'PUT') {
+      return put(url, data, { ...config, signal });
     }
+    return get(url, { ...config, signal });
+  };
+
+  try {
+    const result = await executeQuery(query);
+    return { data: result.data };
   } catch (err) {
     // Handle error of type FetchError
 
     if (isFetchError(err)) {
+      // Attempt auto-refresh on 401 then retry once
+      if (err.status === 401) {
```

Comment:
When an access token expires admin requests will receive a 401 - the first time we see this in the admin we should interpret this as access token expiry and attempt toke renewal to retain admin access

If token renewal fails then we log the user out